### PR TITLE
 feat: add `outputs_validator` to `send_transaction` rpc

### LINKED ISSUE #1866: Add an optional `outputs_validator` parameter to `send_transaction` rpc
**Proposal**

Outputs validator prevents improperly formed transactions from entering the tx-pool, in most cases the leading cause of client sending these transactions is simply because of bug code or misusing SDK, which results in a none-unlockable transaction.

**RPC parameter changes**

```
outputs_validator: Validate the transaction outputs before entering the tx-pool, an optional string parameter, null means using default validator.
```

**Implementation details**

CKB will provide two built-in validators, `default` and `passthrough`.

For `default` validator, these checks should conform to the pseudocode described below

```
transaction.outputs.all{ |output|
    script = output.script
    (script.code_hash == secp256k1_blake160_sighash_all && script.hash_type == "type" && script.args.size == 20) ||
    (script.code_hash == secp256k1_blake160_multisig_all && script.hash_type == "type" && （script.args.size == 20 || (script.args.size == 28 && script.args[20..28].is_valid_since_format))
}
transaction.outputs.all{ |output|
    script = output.type
    script.is_null || script.code_hash == dao && script.hash_type == "type"
}
```

For `passthrough` validator, it will skip validation.

### PR DESCRIPTION
**Pull request by quake**
resolve #1866

### CODE REVIEW

**Comment by TheWaWaR** [comment]
Rebase required.

**Comment by quake** [comment]
> Rebase required.

rebased, please review again

File: rpc/README.md [hunk 1]
```diff
@@ -1950,6 +1950,7 @@ Send new transaction into transaction pool.
 #### Parameters
 
     transaction - The transaction object, struct reference: https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0019-data-structures/0019-data-structures.md#Transaction
+    outputs_validator - Validates the transaction outputs before entering the tx-pool, an optional string parameter (enum: default | passthrough ), null means using default validator, passthrough means skipping outputs validation
```
Comment (keroro520) [inline rpc/README.md]: It worth documenting the behavior of the default validator.

File: rpc/src/module/pool.rs [hunk 2]
```diff
@@ -42,10 +57,23 @@ impl PoolRpcImpl {
...
+    ) -> Result<H256> {
         let tx: packed::Transaction = tx.into();
         let tx: core::TransactionView = tx.into_view();
 
+        if let Err(e) = match outputs_validator {
+            Some(OutputsValidator::Default) | None => {
+                DefaultOutputsValidator::new(self.shared.consensus()).validate(&tx)
+            }
+            Some(OutputsValidator::Passthrough) => Ok(()),
+        } {
```
Comment (keroro520) [inline rpc/src/module/pool.rs]: IMO, returning the error reason would be better for users.

File: rpc/README.md (hunk 1 above)
Comment (quake) [reply rpc/README.md]: added a wiki page: https://github.com/nervosnetwork/ckb/wiki/Transaction-%C2%BB-Default-Outputs-Validator

File: rpc/src/module/pool.rs (hunk 2 above)
Comment (quake) [reply rpc/src/module/pool.rs]: Updated, please review again, thanks.

**Comment by doitian** [comment]
bors r=doitian,keroro520

**Pull request merged**